tests/handlers/test_user_directory.py

```diff
@@ -639,6 +640,64 @@ def _add_user_to_room(
             self.hs.get_storage().persistence.persist_event(event, context)
         )
 
+    def test_making_room_public_doesnt_alter_directory_entry(self):
+        """Per-room names shouldn't go to the directory when the room becomes public.
+
+        I made this a synapse test case rather than a complement one because
+        I think this is (strictly speaking) an implementation choice. Synapse
+        has chosen to only ever use the public profile when responding to a user
+        directory search. There's no privacy leak here, because making the room
+        public discloses the per-room name.
```

Comment:
yeah. This one is a little bit loose. (Maybe we should get spec involved?)

I think Synapse is aiming to do the right thing: just because a room is public (either from history readability or relaxed join rules), doesn't mean you want your name in there to leak (it might be related to the community in use, but unprofessional, personal or weird for use outside of that community* — and sure, a motivated enemy can go looking in all the public rooms everywhere, but we don't necessarily want to make it easy for them).

*a project I know has a public Matrix room where the devs have `(frontend)` or similar labels on their names. But you don't necessarily want that to show up in a personal context.

I have no objections to leaving this as a Synapse test (especially because the knowledge that the room dir changes when a room goes public is kind of Synapse-specific), but really just thinking about preventing this issue for future homeserver implementors.